subsys/dfu/boot/mcuboot.c

```diff
@@ -0,0 +1,254 @@
+{
+	switch (flag) {
+	case BOOT_FLAG_COPY_DONE:
+		*offs = COPY_DONE_OFFS(bank_offs);
+		return 0;
+	case BOOT_FLAG_IMAGE_OK:
+		*offs = IMAGE_OK_OFFS(bank_offs);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int boot_write_flag(int flag, u32_t bank_offs)
+{
+	u8_t buf[FLASH_MIN_WRITE_SIZE];
+	u32_t offs;
+	int rc;
+
+	rc = boot_flag_offs(flag, bank_offs, &offs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	memset(buf, 0xFF, sizeof(buf));
+	buf[0] = BOOT_FLAG_SET;
+
+	flash_write_protection_set(flash_dev, false);
+	rc = flash_write(flash_dev, offs, buf, sizeof(buf));
+	flash_write_protection_set(flash_dev, true);
+
+	return rc;
+}
+
+static int boot_flag_get(int flag, u32_t bank_offs)
+{
+	u32_t offs;
+	int rc;
+	u8_t flag_val;
+
+	rc = boot_flag_offs(flag, bank_offs, &offs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	rc = flash_read(flash_dev, offs, &flag_val, sizeof(flag_val));
+	if (rc != 0) {
+		return rc;
+	}
+
+	return flag_val;
+}
+
+static u8_t boot_copy_done_get(u32_t bank_offs)
+{
+	return boot_flag_get(BOOT_FLAG_COPY_DONE, bank_offs);
+}
+
+static int boot_image_ok_get(u32_t bank_offs)
+{
+	return boot_flag_get(BOOT_FLAG_IMAGE_OK, bank_offs);
+}
+
+static int boot_write_image_ok(u32_t bank_offs)
+{
+	return boot_write_flag(BOOT_FLAG_IMAGE_OK, bank_offs);
+}
+
+static int boot_write_magic(u32_t bank_offs)
+{
+	u32_t offs;
+	int rc;
+
+	offs = MAGIC_OFFS(bank_offs);
+
+	flash_write_protection_set(flash_dev, false);
+	rc = flash_write(flash_dev, offs, boot_img_magic, BOOT_MAGIC_SZ);
+	flash_write_protection_set(flash_dev, true);
+
+	return rc;
+}
+
+static int boot_magic_code_check(const u32_t *magic)
+{
+	int i;
+
+	if (memcmp(magic, boot_img_magic, sizeof(boot_img_magic)) == 0) {
+		return BOOT_MAGIC_GOOD;
+	}
+
+	for (i = 0; i < 4; i++) {
+		if (magic[i] == 0xffffffff) {
+			return BOOT_MAGIC_UNSET;
+		}
+	}
+
+	return BOOT_MAGIC_BAD;
+}
+
+static int boot_magic_state_get(u32_t bank_offs)
+{
+	u32_t magic[4];
+	u32_t offs;
+	int rc;
+
+	offs = MAGIC_OFFS(bank_offs);
+	rc = flash_read(flash_dev, offs, magic, sizeof(magic));
+	if (rc != 0) {
+		return rc;
+	}
+
+	return boot_magic_code_check(magic);
+}
+
+int boot_img_pending_set(int permanent)
+{
+	int rc;
+
+	switch (boot_magic_state_get(FLASH_BANK1_OFFSET)) {
+	case BOOT_MAGIC_GOOD:
+		/* Swap already scheduled. */
+		return 0;
+
+	case BOOT_MAGIC_UNSET:
+		rc = boot_write_magic(FLASH_BANK1_OFFSET);
+		if (rc == 0 && permanent) {
+			rc = boot_write_image_ok(FLASH_BANK1_OFFSET);
+		}
+
+		return rc;
+
+	default:
+		return -EFAULT;
+	}
+}
+
+int boot_img_confirmed_set(void)
+{
+	int rc;
+
+	switch (boot_magic_state_get(FLASH_BANK0_OFFSET)) {
+	case BOOT_MAGIC_GOOD:
+		/* Confirm needed; proceed. */
+		break;
+
+	case BOOT_MAGIC_UNSET:
+		/* Already confirmed. */
+		return 0;
+
+	case BOOT_MAGIC_BAD:
+		/* Unexpected state. */
+		return -EFAULT;
+	}
+
+	if (boot_copy_done_get(FLASH_BANK0_OFFSET) == BOOT_FLAG_UNSET) {
```

Comment:
This is kind of pointless. This code will never be running if the copy isn't completed. Really, none of this code makes sense. It is fine to just write the image_ok and magic value, and be done. There isn't anything the code could do if something were to go wrong anyway. All of this logic just makes it confusing, and isn't going to help anything.